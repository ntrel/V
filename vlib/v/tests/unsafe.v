fn test_ptr_assign() {
    v := 5
    mut p := &v
    p++
    p += 2
    _ := v
}

fn test_ptr_infix() {
    v := 4
    mut q := &v - 1
    q = q + 3
    _ := q
    _ := v
}

struct S1 {}

[unsafe_fn]
fn (s S1) f(){}

fn test_funcs() {
    s := S1{}
    s.f()
}

fn test_c() {
    mut p := C.malloc(4)
    s := 'hope'
    C.memcpy(p, s.str, 4)
}
