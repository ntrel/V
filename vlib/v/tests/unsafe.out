unsafe.v:4:6: warning: pointer arithmetic is only allowed in `unsafe` blocks
    2 |     v := 5
    3 |     mut p := &v
    4 |     p++
      |      ~~
    5 |     p += 2
    6 |     _ := v
unsafe.v:5:7: warning: pointer arithmetic is only allowed in `unsafe` blocks
    3 |     mut p := &v
    4 |     p++
    5 |     p += 2
      |       ~~
    6 |     _ := v
    7 | }
unsafe.v:11:14: warning: pointer arithmetic is only allowed in `unsafe` blocks
    9 | fn test_ptr_infix() {
   10 |     v := 4
   11 |     mut q := &v - 1
      |              ^
   12 |     q = q + 3
   13 |     _ := q
unsafe.v:12:9: warning: pointer arithmetic is only allowed in `unsafe` blocks
   10 |     v := 4
   11 |     mut q := &v - 1
   12 |     q = q + 3
      |         ^
   13 |     _ := q
   14 |     _ := v
unsafe.v:24:7: warning: method `S1.f` must be called from an `unsafe` block
   22 | fn test_funcs() {
   23 |     s := S1{}
   24 |     s.f()
      |       ~~~
   25 | }
   26 |
unsafe.v:28:16: warning: function `C.malloc` must be called from an `unsafe` block
   26 |
   27 | fn test_c() {
   28 |     mut p := C.malloc(4)
      |                ~~~~~~~~~
   29 |     s := 'hope'
   30 |     C.memcpy(p, s.str, 4)
unsafe.v:30:7: warning: function `C.memcpy` must be called from an `unsafe` block
   28 |     mut p := C.malloc(4)
   29 |     s := 'hope'
   30 |     C.memcpy(p, s.str, 4)
      |       ~~~~~~~~~~~~~~~~~~~
   31 | }
